text/0000-language-semver.md

```diff
@@ -0,0 +1,494 @@
+
+Opt-in changes allow us to greatly expand the scope of the kinds of
+additions we can make without breaking existing code, but they are not
+applicable in all situations. A good rule of thumb is that an opt-in
+change is only appropriate if the exact effect of the older code can
+be easily recreated in the newer system with only surface changes to
+the syntax.
+
+Another view is that opt-in changes are appropriate if those changes
+do not affect the "abstract AST" of your Rust program. In other words,
+existing Rust syntax is just a serialization of a more idealized view
+of the syntax, in which there are no conflicts between keywords and
+identifiers, syntactic sugar is expanded, and so forth. Opt-in changes
+might affect the translation into this abstract AST, but should not
+affect the semantics of the AST itself at a deeper level. This concept
+of an idealized AST is analagous to the "elaborated syntax" described
+in [RFC 1105], except that it is at a conceptual level.
+
+So, for example, the conflict between new keywords and existing
+identifiers can (generally) be trivially worked around by renaming
+identifiers, though the question of public identifiers is an
+interesting one (contextual keywords may suffice, or else perhaps some
+kind of escaping syntax -- we defer this question here for a later
+RFC).
+
+In the previous section on breaking changes, we identified various
+criteria that can be used to decide how to approach a breaking change
+(i.e., how far to go in attempting to mitigate the fallout). For the
+most part, those same criteria also apply when deciding whether to
+accept an "opt-in" change:
+
+- How many crates on `crates.io` would break if they "opted-in" to the
+  change, and would opting in require extensive changes?
+- Does the change silently change the result of running the program,
+  or simply cause additional compilation failures?
+  - Opt-in changes that silently change the result of running the
+    program are particularly unlikely to be accepted.
+- What changes are needed to get code compiling again? Are those
+  changes obvious from the error message?
+
+# Drawbacks
+
+**Allowing unsafe code to continue compiling -- even with warnings --
+raises the probability that people experiences crashes and other
+undesirable effects while using Rust.** However, in practice, most
+unsafety hazards are more theoretical than practical: consider the
+problem with the `thread::scoped` API. To actually create a data-race,
+one had to place the guard into an `Rc` cycle, which would be quite
+unusual. Therefore, a compromise path that warns about bad content but
+provides an option for gradual migration seems preferable.
+
+**Deprecation implies that a maintenance burden.** For library APIs,
+this is relatively simple, but for type-system changes it can be quite
+onerous. We may want to consider a policy for dropping older,
+deprecated type-system rules after some time, as discussed in the
+section on *unresolved questions*.
```

Comment:
there's also a testing burden being introduced here, right?

Or at least, we've opened ourselves up to the potential for tricky-to-diagnose bugs that depend on the combination of using one actual `rustc` version being passed another language version to emulate.